---
title: 'Biology 351: Lab Week 10 S24'
output:
  pdf_document: default
  html_notebook: default
---

*Introduction*

We are back to R to analyze and plot our allele frequency data that we generated over the past several weeks of class. Before spring break, you created a filtered vcf file using just your two samples; you'll use this next week to do an FST analysis looking for genetic differences between your two populations. Over spring break, I created a large vcf file that includes all of the samples--44 different pooled DNA samples from the past 6 years in Virginia and from all over the east coast. There are over 6,000,000 SNPs in this file! This week we are going to use these data to run and  a principal component analysis (PCA) and do a preliminary comparison of FST across all the samples before moving on to do a genome-wide scan for FST between our focal samples next week. 

To start, I have already turned the data encoded in the AD and RD spaces of the vcf into an R data table.

**Question:** (From a few weeks ago): What do AD and RD stand for? What information do they hold?
**Answer here:**

**Question** How did I convert RD and AD values into an allele frequency? 
**Answer here**

We need to use some Linux code to get some large files off of `spydur` and to the Rstudio server where we are working today.  We will use a Linux function called `scp` (short for "secure copy") to do that. It will copy the named files to your current working directory on the Rstudio server (indicated by the `./` at the end of the command). We can run Linux commands through Rstudio by putting `{bash}` at the top of the code. 

Replace `yournetid` after the `scp` with your actual netid below in each line of code and run this bash code chunk.  You should see these files appear in your working folder to the right. The first line sets up some information that will allow us to use one of the R packages that we'll be using today. 


```{bash}
export LD_LIBRARY_PATH=/usr/lib64:/usr/local/sw/anaconda3/lib:/usr/pgsql-13/lib
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/sw/anaconda3/lib
scp yournetid@spydur:~/shared_perickso/ZP_chr1.vcf ./
scp yournetid@spydur:~/shared_perickso/ZP_chr3.vcf ./

```

*Part 1: Principal Components Analysis (PCA)*

First, let's load the packages and data in. I've already been processed the allele frequencies into a data.table for each pooled sample, but it's a big file and will take ~30 seconds to load. It will load in as an object called `y`.
 
```{r}
library(data.table)
library(foreach)
library(ggplot2)

#read in vcf allele frequency data. Will load in as a data.table called freqs.wide
load("downsampled_allele_freqs_S24.Rdata")
```

When we do our analysis we want to remove SNPs  that are closely linked in haplotypes. These SNPs will tend to have correlated allele frequencies and may bias our analysis. I ran a short program that will keep only one SNP in every 2000 bp window of the genome. That way, any two SNPs in the analysis are, on average, about 2000 bp apart, which is far enough to remove most genetically linked SNPs and have one SNP representing each haplotype. 

**Question** Look at the `y` table. How are the data organized?

**Answer here:**


**Question** Each entry in y is representing data for a specific sample at a specific location. The numbers vary between 0 and 1. What are these numbers?

**Answer here:**


**Question** How many SNPs are currently in our dataset? You can use the functions `dim()` or `nrow()` to find out. 

**Answer here:**

First let's do a principal components analysis using our  data for the entire genome. Note that while most of the PCAs we have looked at in class have used *genotypes of individuals* at many SNPs to calculate PCA, we can also use *allele frequencies of populations* for a similar calculation. 

Now we'll do the actual principal components calculation. R has a built-in function called `prcomp` to do this. You can run the whole chunk of code below to execute the PCA; we won't worry about the specific functions used here but if you have questions I'm happy to explain how this code works. 

```{r}
#we only want to use SNPs with no missing data. This function will remove any rows of the matrix with missing data
pca.matrix<-as.matrix(y[,4:47]) 
#this assigns marker names to each row of the matrix
  rownames(pca.matrix)=y$marker  
  #omit missing data
pca.matrix.complete<-na.omit(pca.matrix)
  #calculate PCA with prcomp() function
  pca.results<-prcomp(pca.matrix.complete, scale=T)
  #turn PCA results into a data.table
  pca.data<-as.data.table(pca.results$rotation)
  #add ID columns to that table
  pca.data[,Pool:=rownames(pca.results$rotation)]
  #look at just PC1 through PC4
  pca.data<-pca.data[,.(Pool, PC1, PC2, PC3, PC4)]

```
**Question** Explain in your own words what a PCA does and what data will be the raw input to this PCA.

**Answer here**

**Question** Most of the PCAs we've seen in class have compared genotypes from individual organisms. How is this PCA different? 

**Answer here**

**Question** Look at the `pca.data` object. What does it have in the rows and columns? 

**Answer here**

**Question** Based on what you know about principle components and how these samples were generated and processed,  what does it mean if two of our samples samples have similar values of PC1 and PC2? Be specific in your answer.

**Answer here**

To make an informative plot, we need to bring in some information about each sample so we know what we are looking at and can color-code our points appropriately. We'll read in a file that has information about each sample in the code below and save it as a data.table called `info`.

```{r}
info<-fread("samples.txt")
```

**Question** Look at the `info` table you just loaded. How will this information be useful in analyzing the PCA?

**Answer here**

**Question** To combine these two data tables together, we need a column that has the same name and same information in both tables to match up the data. What is the name of the column that is shared in common with `pca.data` and `info` ?

**Answer here**


Now set up the command below to merge the sample and pca data by replacing `"name.of.shared.column"` with the column you identified above. Keep the quotes around the column name. Note that this command takes the f two data tables and the column they have in common, then uses a shared column to match up the data. 

```{r}
pca.data<-merge(pca.data, info, by="name.of.shared.column") 
 

```

**Question** Now look at your pca.data table. Describe what the merge function just did.

**Answer here**

Now let's make some plots of our whole-genome PCA data. First make a plot comparing the samples by **collection location** by completing the x, y, and color information below to make a standard PCA plot as we have typically seen them. You will need to put the name of a column from pca.data in for the x, y, and color. 

```{r}
ggplot(pca.data)+geom_point(aes(x=   , 
                            y=  , 
                            color=    )) 

```

**Question** Do you see any patterns  based on the location of the samples? 
**Answer here**

You can also plot only some of your data by using conditional statements to extract only some rows. Here the conditional statement is `[`Year==2021]`, which means that only the rows of `pca.data` that have 2021 in the `Year` column will be plotted. You may recall that conditional statements require two  == signs to identify only rows that have a particular property. Because Year is a numeric column, we do not need quotes around the 2021. 

```{r}
ggplot(pca.data[Year==2021])+geom_point(aes(x= PC1, 
                            y= PC2, 
                            color= Location)) 
```
**Question** Do the samples appear to cluster by location when looking at only one year?
**Answer here**


Next, let's make a plot that will specifically highlight the two samples you'll be comparing for your analysis so you can compare them to all the other samples in the PCA. We will first need to make a new column that identifies the samples you worked on. In this column, we will put the word "TRUE" for the two rows that are your samples, and all the other rows will say "FALSE". You'll need this plot for your final presentation so make sure you get this part done correctly and feel free to ask questions if you need help

```{r}
#first make a new column called "focal_sample" that will be filled with "FALSE" values (every row will say "FALSE" for this column). You don't need to change anything about this line of code
pca.data[, focal_sample:=FALSE]

#now, for the rows that contain your two samples, change the "your.sample.name" to the names of your two samples (ZP_##). Because the Pool column contains character data, the name you are matching has to go in quotes.

pca.data[ Pool == "your.sample.name1" , focal_sample:=TRUE] #complete this line by changing your.sample.name1
pca.data[ Pool == "your.sample.name2" , focal_sample:=TRUE] #complete this line by changing your.sample.name2=

#print pca.data to make sure this worked; if not ask me for help
pca.data
```

Now you can use your new column to make a plot to identify your samples in the whole dataset. You'll want to color code the plot with the "focal sample" column so that your points will be a different color than all the others. 

```{r}
ggplot(pca.data)+geom_point(aes(x=  , 
                            y=   ,
                            color= ))
```

**Question** Compared to the overall variation in the samples, how similar are your two samples that you'll be comparing? Do they differ in PC1, PC2, or both?

**Answer here**

We were focusing on the whole genome, but previous research on Zaprionus has suggested that there are some interesting trends of population differentiation with individual chromosomes. Let's repeat our PCA, but now we will calculate the PCA for each chromosome separately and compare the results across chromosomes. This will allow us to see if certain chromosomes have stronger genetic differences than others. Like a few weeks ago, we'll use a "for loop" that processes the chromosomes one at a time. The mechanics of how this code works are too complicated to go through today, so you can just run this code and proceed to the steps below. 

```{r}
pca.chrom<-foreach(chrom=c(1:5))%do%{
  print(chrom)
  #this line will pull out only the data for one chromosome
  y.chrom<-y[chr==chrom,3:47]
  pca.matrix.chrom<-as.matrix(y.chrom)
#this assigns marker names to each row of the matrix
  rownames(pca.matrix.chrom)=y.chrom$marker
#only want SNPs with calls in all pops
  pca.matrix.chrom.complete<-na.omit(pca.matrix.chrom)
  #calculate PCA for the chromosome
  pca.results.chrom<-prcomp(pca.matrix.chrom.complete, scale=T)
  #turn it into a data table
  pca.data.chrom<-as.data.table(pca.results.chrom$rotation)
  #add names of pools
  pca.data.chrom[,Pool:=rownames(pca.results.chrom$rotation)]
  pca.data.chrom[,chrom:=chrom]
  return(pca.data.chrom)
}
pca.chrom<-rbindlist(pca.chrom)
pca.chrom<-pca.chrom[,.(Pool,chrom, PC1, PC2, PC3, PC4)]

```
**Question** Which line of the code above caused the analysis to be restricted to just one chromosome at a time?

**Answer here**

Now use the `merge` command that you learned above to merge your new `pca.chrom` object with the sample information table, `info`. Make sure you change all the variables that need to be changed to execute this command.

```{r}
#merge here using similar command to above (or see cheat sheet)


#confirm that your merge worked


```

Now we'll make the same plots we made above, color coded by location, but plot each chromosome separately in "facets" with the `facet_wrap` command in `ggplot`. To add facets you add `+ facet_wrap(~variable.to.facet.by, scales="free")` to the final line of your plotting command. You'll need to change `variable.to.facet.by` to the name of the variable that you want to split up your plots with. Use your code from above as an example to make a new plot that is color-coded by Location and faceted by chromosome so that there is a different plot for each chromosome. 

```{r}
#write plotting code here

ggplot(pca.chrom)+geom_point(aes(x=   , 
                            y=  , 
                            color=    )) + facet_wrap(~, scales="free")


```


**Question** How do the individual chromosome results compare to the full genome results? What could explain the differences? Do any chromosomes stand out? 

**Answer here**

*Part 2: Independent investigation*

You've now seen how to make PCA plots that color coded, and you've seen how to subset data (use only some of the data available with conditional statements, like [Year==2021] or [Season=="early"]) and how to split your data into multiple graphs with facet_wrap(). Your challenge now is to use the various columns available in the sample information to come up with another way to plot the data to look for patterns in the PCA. You could look at a single orchard over time, you could compare seasons, you could compare multiple orchards in one year,  you could look at multiple locations within a year, etc. You can use the whole-genome data or the individual chromosome data. Below are the metadata describing the different columns to help you.

    Pool: The name of the sequencing library.
    Location: The location flies were collected:
        CM = Carter Mountain, Charlottesville, VA
        FL = locations in anad around, Miami, FL
        HPO = Hanover Peach Orchard, Richmond, VA
        GA = Hillcrest Orchard, Ellijay, GA
        CK = Cross Keys Orchard, Harrisonburg, VA
        MA = Acton, MA (flies combined from multiple orchards)
        CT = Lyman Orchards, Middleton, CT
        PA = Linvilla Orchards, Media, PA
        CH = Chiles Orchards, Crozet, VA
    Experiment: Whether the sample was part of the seasonal sampling or latitudinal experiment
    State: The location the samples were collected
    Year: The year of sampling
    Month: The calendar month of sampling
    Season: whether the sample was collected early (July-August), middle (September), or late (October-November) during hte Z. indianus growing season
    Sample Date: The day flies were collected
    females: The number of female flies in the pool
    males: The number of male flies in the pool
    total: The total number of flies in the pool

**Question** What is the PCA pattern you are interested in looking at and why? What question will this answer?
**Answer**

**Question** How will you set up your analysis? What data will you use, how will you color code, and will you facet your graph?
**Answer**

Before moving on, check with Dr. E about your answers. Then, develop the code for your plot below and make your plot. 
```{r}



```

At the end of the lab you will be asked to write a brief figure legend and results statement for this analysis. I recommend saving that until the end of the lab so you have enough time to get help with any additional coding questions before the end of the lab period.

*Part 3: Genome-wide FST between all pairs of samples*

For this part we are going to use a package called `poolfstat`. This package is specifically designed to do FST and other population genetic tests on pooled population sequencing data like we have. 

To speed things up, we will just be conducting this analysis on a small region of the genome on chromosome 3.
Run the code below to read our vcf file into R with poolfstat. We wont' worry about the mechanics of this code; just run it.

```{r}
library(poolfstat)

sample.names=fread("vcf_sample_names.txt", header =F)$V1
setkey(info, Pool)
pool.sizes=info[sample.names, total]
vcf.dat<-vcf2pooldata(vcf.file="ZP_chr3.vcf",
                  poolnames=sample.names,
                  poolsizes=pool.sizes,
                  min.cov.per.pool=25)
```
One of the functions of this package can calculate "pairwise" FST using each sample. 

**Question** What is the typical scale of FST values, and what do higher and lower FST values mean ?
**Answer here**

**Question:** What do you think "pairwise' FST means? 
**Answer here**

Run the code below to generate the pairwise FST values.

```{r}
pfst<-compute.pairwiseFST(vcf.dat,verbose=TRUE) 
```

Let's look at the `pfst` object created in our last command. It created a bunch of data, but the actual FST values are saved as something called `PairwiseFSTmatrix`, the first 10 rows and columns of which will be printed with the command below

```{r}
pfst@PairwiseFSTmatrix[1:10, 1:10]
```

**Question** What do you think is shown in these data? 
**Answer here**


**Question** Why are there "NA"s along a diagonal line in this table?
**Answer here**


**Question** Typically we've seen FST calculated for each, but here we used ~7,000 SNPs and got out a single number. What do you think this number represents?
**Answer here**



The poolfstat package includes a built-in function to create a heatmap of the FST values for each populations comparison. Heatmaps use different colors to indicate different numerical values. Run the code below to generate the default heatmap from the program.

```{r}
heatmap(pfst,symm=T)

```
**Question** What do you think this heatmap is attempting to illustrate? What are the rows, columns, and colors representing?
**Answer here**


**Question** I don't know about you, but I find this plot pretty hard to interpret. What are some things that are wrong with this plot? You should be able to identify at least 2-3 issues.
**Answer here**


Let's make a better heatmap so that we can be certain of what we are looking at. You'll need to run the first three lines, then look at the new object you've created and update the plotting information in the second chunk of code to actually  make the plot.

```{r}
#first we melt our data into a long format
library(reshape2)
library(viridis)
fst.melt<-as.data.table(melt(pfst@PairwiseFSTmatrix))
names(fst.melt)=c("Sample1", "Sample2", "FST")
fst.melt
```

Now look at the FST.melt object you printed above. You'll need to understand how it's laid out to update the code for your heatmap below.

```{r}
#now update the code to make a customize heatmap plot
ggplot(fst.melt)+
  geom_tile(aes(x=  , #what column makes the x axis? 
                y=  , #what column makes the y axis?  
                fill=   ))+ #what column should determine the coloring?  
  #this next line is going to rotate the angle of the letters on teh x axis
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))+
  #this line removes x and y axis labels and labels the color scale
  labs(x=NULL, y=NULL, fill="FST")+
  #this line will use prettier colors!
  scale_fill_viridis()

```
**Question** Which samples appear to be the most different from all the other samples? Based on the `info` table, what do they have in common? 
**Answer here**

**Question** Does this result support or contradict the chromosome-specific PCA analysis?
**Answer here**

We just looked at the region of the genome tha thas some of the highest differentiation between samples. Now we are going to quickly repeat everything we just did but for a different, randomly chosen region of the genome (somewhere in the middle of chromosome 1) for comparison's sake. Run the whole chunk of code below, you don't have to change anything. It will take a few seconds to run.
```{r}
vcf.dat<-vcf2pooldata(vcf.file="ZP_chr1.vcf",
                  poolnames=sample.names,
                  poolsizes=pool.sizes,
                  min.cov.per.pool=25)

pfst<-compute.pairwiseFST(vcf.dat,verbose=TRUE) 
fst.melt<-melt(pfst@PairwiseFSTmatrix)
ggplot(fst.melt)+
  geom_tile(aes(x= Var1 , 
                y= Var2 , 
                fill= value  ))+ 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))+
  labs(x=NULL, y=NULL, fill="FST")+
  #this line will use prettier colors!
  scale_fill_viridis()
```
**Question** Looking at the FST scale, do there seem to be a lot of differences between samples in this region of the genome? Does that agree with the chromosome PCA analysis?   
**Answer here**

**Question** Find the point that is the intersection of your two samples--How does the FST between your two samples compare to the range of values in all the comparisons? Relatively low? High? 
**Answer here**

**Final task**For the independent analysis that you conducted in part 2, write a figure legend, including a figure title, and a short results statement describing the PCA results and citing the figure. Remember that your figure legend should allow the reader to understand what was done to produce the data and interpret the data without looking at additional information. So you need to explain where the samples came from, how they were analyzed (briefly), and what any labels/colors/etc mean. The results paragraph should orient the reader to purpose of the analysis and then describe key results, referring back to the figure (which you can name "Figure 1")

**Figure legend here**

**Results here**

When you are done, please use Preview-->Knit to HTML to generate a nice preview, then save it as a PDF and upload to Blackboard. Make sure the PDF includes all of your plots! 